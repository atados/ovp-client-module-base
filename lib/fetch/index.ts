const isServer = typeof window === 'undefined'
import { Options } from './fetch.client'

const polyfilledFetch = isServer
  ? // tslint:disable-next-line:no-var-requires
    require('./fetch.server')
  : // tslint:disable-next-line:no-var-requires
    require('./fetch.client')

type FetchFunction = <Payload = any>(
  url: string,
  options?: Options,
) => Promise<Payload>
export const fetch = polyfilledFetch.fetch as FetchFunction
export const fetchJSON = polyfilledFetch.fetchJSON as FetchFunction
export const fetchAPI = polyfilledFetch.fetchAPI as FetchFunction
