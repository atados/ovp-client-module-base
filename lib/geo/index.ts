const GeoModule =
  typeof window === 'undefined'
    ? // tslint:disable-next-line:no-var-requires
      require('./geo.server')
    : // tslint:disable-next-line:no-var-requires
      require('./geo.client')

export const createGeolocationObject = GeoModule.createGeolocationObject
