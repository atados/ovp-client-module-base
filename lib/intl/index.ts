import hoistNonReactStatics from 'hoist-non-react-statics'
import { injectIntl } from 'react-intl'

export const hoistStatics = <T = any>(
  higherOrderComponent: (
    BaseComponent: React.ComponentType<T>,
  ) => React.ComponentType<T>,
) => (BaseComponent: React.ComponentType<T>) => {
  const NewComponent = higherOrderComponent(BaseComponent)
  hoistNonReactStatics(NewComponent, BaseComponent)

  return NewComponent
}

// @ts-ignore
export const withIntl = hoistStatics(injectIntl)
