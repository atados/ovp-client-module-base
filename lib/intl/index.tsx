import useIntl from '~/hooks/use-intl'
import { InjectedIntlProps } from 'react-intl'

export function withIntl<Props extends InjectedIntlProps>(
  Component: React.ComponentType<Props>,
): React.FC<Omit<Props, keyof InjectedIntlProps>> {
  return props => {
    const intl = useIntl()
    return <Component {...(props as Props)} intl={intl} />
  }
}
