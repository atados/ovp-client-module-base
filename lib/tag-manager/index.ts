type DataLayerEvent = string | number | boolean | null | Date
export interface DataLayerMessage {
  [variableName: string]: DataLayerEvent
}

declare global {
  interface Window {
    dataLayer: DataLayerMessage[]
  }
}

export function setupDataLayer(gtmId: string) {
  window.dataLayer = window.dataLayer || []
  pushToDataLayer({
    js: new Date(),
    config: gtmId,
  })
}

const isBrowser = typeof window !== 'undefined'
export function pushToDataLayer(event: DataLayerMessage) {
  if (isBrowser && window.dataLayer) {
    window.dataLayer.push(event)
  }
}
