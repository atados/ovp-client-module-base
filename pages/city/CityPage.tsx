import React from 'react'
import { NextPage } from 'next'
import { useRouter } from 'next/router'
import { FormattedMessage } from 'react-intl'
import { mountAPIPathToSearchProjects } from '~/pages/city/utils'
import { ApiPagination } from '~/redux/ducks/search'
import ProjectCard from '~/components/ProjectCard'
import { RootState } from '~/redux/root-reducer'
import useFetchAPI from '~/hooks/use-fetch-api'
import { useSelector } from 'react-redux'
import Layout from '~/components/Layout'
import Icon from '~/components/Icon'
import Meta from '~/components/Meta'
import { API } from '~/types/api'
import BannerGradientOverlay from '~/components/Styled/BannerGradientOverlay'
interface CityProps {
  cityName: string
  selectedCauseId?: string
  selectedSkillId?: string
}

const reduxStateSelector = (state: RootState) => state.startup

const City: NextPage<CityProps> = ({
  cityName,
  selectedCauseId,
  selectedSkillId,
}) => {
  const router = useRouter()
  const { causes, skills } = useSelector(reduxStateSelector)
  const response = useFetchAPI<ApiPagination<API.Project>>(
    mountAPIPathToSearchProjects(cityName, selectedCauseId, selectedSkillId),
  )
  const handleInputChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const queryObject = {
      cause: selectedCauseId,
      skill: selectedSkillId,
      [event.target.name]: event.target.value,
    }

    router.push({
      pathname: router.pathname.replace('[cityName]', cityName),
      query: queryObject,
    })
  }

  const projects = response.data?.results || []

  return (
    <Layout toolbarProps={{ float: true, flat: true, className: 'bg-none' }}>
      <Meta />
      <span className="px-md-3"></span>
      <BannerGradientOverlay
        className="p-toolbar bg-cover bg-center mb-10"
        style={{ backgroundImage: `url('/static/banners/home-fast.jpg')` }}
      >
        <div className="relative z-50">
          <div className="container px-3 pt-8">
            <div className="w-16 h-16 rounded-full bg-white mx-auto mb-4 flex justify-center items-center">
              <Icon name="place" className="text-4xl text-blue-600" />
            </div>
            <h1 className="text-white text-6xl font-medium text-center">
              {cityName}
            </h1>
            <p className="max-w-xl text-xl text-center text-white mx-auto mb-6">
              <FormattedMessage
                id="home.subtitle"
                defaultMessage="O Atados é uma plataforma de voluntariado que conecta pessoas a oportunidades voluntárias."
              />
            </p>
            <div className="bg-white rounded-lg max-w-3xl mx-auto mb-4 p-2 md:flex">
              <select
                value={selectedCauseId}
                className="rounded-lg input md:mr-2 block mb-2 md:mb-0 bg-gray-300 border-0 text-xl px-4 h-12"
                name="cause"
                onChange={handleInputChange}
                role="cause-filter"
              >
                <option key={0} value="">
                  Filtre por causa
                </option>
                {causes.map(cause => (
                  <option key={cause.id} value={cause.id}>
                    {cause.name}
                  </option>
                ))}
              </select>
              <select
                value={selectedSkillId}
                className="rounded-lg input md:mr-2 block mb-2 md:mb-0 bg-gray-300 border-0 text-xl px-4 h-12"
                name="skill"
                onChange={handleInputChange}
                role="skill-filter"
              >
                <option key={0} value="">
                  Filtre por habilidade
                </option>
                {skills.map(skill => (
                  <option key={skill.id} value={skill.id}>
                    {skill.name}
                  </option>
                ))}
              </select>
            </div>
          </div>
          <svg
            xmlns="http://www.w3.org/2000/svg"
            width="100%"
            height="96px"
            viewBox="0 0 100 100"
            version="1.1"
            preserveAspectRatio="none"
            className="block h-16"
          >
            <path
              fill="#fff"
              d="M0,0 C16.6666667,66 33.3333333,99 50,99 C66.6666667,99 83.3333333,66 100,0 L100,100 L0,100 L0,0 Z"
            ></path>
          </svg>
        </div>
      </BannerGradientOverlay>
      <div className="container mx-auto px-2">
        <h4 className="mb-4">Vagas recentes</h4>
        <div className="flex -mx-2 flex-wrap">
          {projects.map(project => (
            <div
              key={project.slug}
              className="px-2 w-full sm:w-1/2 lg:w-1/4 mb-6"
            >
              <ProjectCard {...project} />
            </div>
          ))}
        </div>
      </div>
    </Layout>
  )
}

City.getInitialProps = async context => {
  return {
    cityName: String(context.query.cityName),
    selectedCauseId: context.query.cause && String(context.query.cause),
    selectedSkillId: context.query.skill && String(context.query.skill),
  }
}

export default City
