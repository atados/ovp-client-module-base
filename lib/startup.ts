import { colors, dev } from '~/common/constants'
import { fetchAPI } from '~/lib/fetch'
import { API } from '~/base/types/api'

export async function getStartupData(
  locale?: string,
): Promise<API.StartupPayload> {
  try {
    const {
      causes,
      skills,
      volunteer_count: volunteerCount,
      nonprofit_count: nonprofitCount,
    } = await fetchAPI<{
      volunteer_count: number
      nonprofit_count: number
      causes: API.Cause[]
      skills: API.Skill[]
    }>('/startup/', {
      headers: {
        'Accept-Language': locale,
      },
    })
    return {
      skills,
      causes: causes
        .map((cause, i) => ({
          ...cause,
          color: colors[i % (colors.length - 1)],
        }))
        .sort((x, y) => x.id - y.id),
      stats: {
        volunteers: volunteerCount || 0,
        organizations: nonprofitCount || 0,
      },
    }
  } catch (error) {
    if (dev) {
      console.error('> Failed to fetch startup')
    } else {
      console.error(error)
    }
  }

  return {
    skills: [],
    causes: [],
    stats: {
      volunteers: 0,
      organizations: 0,
    },
  }
}
