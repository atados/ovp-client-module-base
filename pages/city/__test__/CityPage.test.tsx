import React from 'react'
import renderer from 'react-test-renderer'
import { RouterContext } from 'next/dist/next-server/lib/router-context'
import CityPage from '..'
import { render, fireEvent, wait } from '@testing-library/react'

jest.mock('~/components/Layout', () => props => <div>{props.children}</div>)

jest.mock('react-redux', () => ({
  useSelector: () => ({
    causes: [
      { id: 1, name: 'cause1' },
      { id: 2, name: 'cause2' },
    ],
    skills: [
      { id: 1, name: 'skill1' },
      { id: 2, name: 'skill2' },
    ],
  }),
}))

jest.mock('~/hooks/use-fetch-api', () => () => ({
  data: {
    results: [
      {
        id: 1,
        name: 'Project 1',
        slug: 'project_1',
      },
      {
        id: 2,
        name: 'Project 2',
        slug: 'project_2',
      },
    ],
  },
}))

jest.mock('~/components/ProjectCard', () => props => (
  <div {...props}>~/components/ProjectCard - {props.children}</div>
))

describe('City Page', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<CityPage cityName="Sao Paulo" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('change the route when the filters change', async () => {
    const router: any = {
      pathname: '/cidades/[cityName]',
      push: jest.fn(),
    }
    const { getByRole } = render(
      <RouterContext.Provider value={router}>
        <CityPage cityName="São Paulo" selectedSkillId="2" />
      </RouterContext.Provider>,
    )
    fireEvent.change(getByRole('cause-filter'), {
      target: { name: 'cause', value: '1' },
    })
    await wait()
    expect(router.push).toBeCalledWith({
      pathname: '/cidades/São Paulo',
      query: { cause: '1', skill: '2' },
    })
  })

  it('extract initial props correctly', async () => {
    expect(
      await CityPage.getInitialProps!({
        query: {
          cityName: 'SP',
          cause: 1,
        },
      } as any),
    ).toMatchObject({
      cityName: 'SP',
      selectedCauseId: '1',
    })
    expect(
      await CityPage.getInitialProps!({
        query: {
          cityName: 'SP',
          skill: 1,
        },
      } as any),
    ).toMatchObject({
      cityName: 'SP',
      selectedSkillId: '1',
    })
    expect(
      await CityPage.getInitialProps!({
        query: {
          cityName: 'SP',
          cause: 1,
          skill: 1,
        },
      } as any),
    ).toMatchObject({
      cityName: 'SP',
      selectedCauseId: '1',
      selectedSkillId: '1',
    })
  })
})
