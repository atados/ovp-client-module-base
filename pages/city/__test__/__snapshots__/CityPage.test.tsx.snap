// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`City Page renders correctly 1`] = `
.c0 {
  position: relative;
}

.c0::after {
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  opacity: 0.6;
  background: linear-gradient( 180deg,#00355b,#4d9fda );
}

<div>
  <span
    className="px-md-3"
  />
  <div
    className="c0 p-toolbar bg-cover bg-center mb-10"
    style={
      Object {
        "backgroundImage": "url('/static/banners/home-fast.jpg')",
      }
    }
  >
    <div
      className="relative z-50"
    >
      <div
        className="container px-3 pt-8"
      >
        <div
          className="w-16 h-16 rounded-full bg-white mx-auto mb-4 flex justify-center items-center"
        >
          <span
            className="icon im text-4xl text-blue-600"
          >
            place
          </span>
        </div>
        <h1
          className="text-white text-6xl font-medium text-center"
        >
          Sao Paulo
        </h1>
        <p
          className="max-w-xl text-xl text-center text-white mx-auto mb-6"
        >
          FormattedMessage
        </p>
        <div
          className="bg-white rounded-lg max-w-3xl mx-auto mb-4 p-2 md:flex"
        >
          <select
            className="rounded-lg input md:mr-2 block mb-2 md:mb-0 bg-gray-300 border-0 text-xl px-4 h-12 "
            disabled={false}
            name="cause"
            onChange={[Function]}
            role="cause-filter"
          >
            <option
              value=""
            >
              Filtre por causa
            </option>
            <option
              value={1}
            >
              cause1
            </option>
            <option
              value={2}
            >
              cause2
            </option>
          </select>
          <select
            className="rounded-lg input md:mr-2 block mb-2 md:mb-0 bg-gray-300 border-0 text-xl px-4 h-12 "
            disabled={false}
            name="skill"
            onChange={[Function]}
            role="skill-filter"
          >
            <option
              value=""
            >
              Filtre por habilidade
            </option>
            <option
              value={1}
            >
              skill1
            </option>
            <option
              value={2}
            >
              skill2
            </option>
          </select>
        </div>
      </div>
      <svg
        className="block h-16"
        height="96px"
        preserveAspectRatio="none"
        version="1.1"
        viewBox="0 0 100 100"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0,0 C16.6666667,66 33.3333333,99 50,99 C66.6666667,99 83.3333333,66 100,0 L100,100 L0,100 L0,0 Z"
          fill="#fff"
        />
      </svg>
    </div>
  </div>
  <div
    className="container mx-auto px-2"
  >
    <h4
      className="mb-4"
    >
      Vagas recentes
    </h4>
    <div
      className="flex -mx-2 flex-wrap"
    >
      <div
        className="px-2 w-full sm:w-1/2 lg:w-1/4 mb-6"
      >
        <div
          id={1}
          name="Project 1"
          slug="project_1"
        >
          ~/components/ProjectCard - 
        </div>
      </div>
      <div
        className="px-2 w-full sm:w-1/2 lg:w-1/4 mb-6"
      >
        <div
          id={2}
          name="Project 2"
          slug="project_2"
        >
          ~/components/ProjectCard - 
        </div>
      </div>
    </div>
  </div>
</div>
`;
